connectors/bitbucket-server/src/main/java/com/vmware/connectors/bitbucket/server/BitbucketServerController.java

```diff
@@ -0,0 +1,483 @@
+        final HttpHeaders headers = buildHeaders(authHeader);
+
+        // Get current version of the pull request. Pull request "version" changes when we do any actions on it.
+        // When the pull request is raised, the current value will be 0.
+        // For example, when we approve the pull request, then the version will change from 0 to 1.
+        // We have to add the latest version of the pull request URI to do any actions. Otherwise, the ACTION will be rejected.
+        // If the build for the branch is going on, then the actions would be rejected.
+        // final Single<String> version = getVersion(headers, baseUrl, pullRequest);
+        return getVersion(headers, baseUrl, pullRequest)
+                .flatMap(version -> {
+
+                    final URI uri = UriComponentsBuilder.fromHttpUrl(baseUrl + "/rest/api/1.0/projects/{PROJECT_KEY}/repos/{REPOSITORY_SLUG}/pull-requests/{PULL_REQUEST_ID}/{ACTION}")
+                            .queryParam("version", version)
+                            .buildAndExpand(
+                                    ImmutableMap.of(
+                                            PROJECT_KEY, pullRequest.getProjectKey(),
+                                            REPOSITORY_SLUG, pullRequest.getRepositorySlug(),
+                                            PULL_REQUEST_ID, pullRequest.getPullRequestId(),
+                                            ACTION, bitBucketServerAction.getAction()
+                                    )
+                            )
+                            .toUri();
+
+                    // This header is added for skipping the CSRF check. Otherwise, Atlassian denies the pull request ACTION.
+                    headers.add(ATLASSIAN_TOKEN, "no-check");
+
+                    final ListenableFuture<ResponseEntity<String>> response = this.rest.exchange(
+                            uri,
+                            HttpMethod.POST,
+                            new HttpEntity<>(headers),
+                            String.class);
+
+                    return Async.toSingle(response);
+                });
+
+    }
+
+    private HttpHeaders buildHeaders(final String authHeader) {
+        final HttpHeaders headers = new HttpHeaders();
+        headers.add(AUTHORIZATION, authHeader);
+        return headers;
+    }
+
+    private Single<String> getVersion(final HttpHeaders headers,
+                              final String baseUrl,
+                              final BitbucketServerPullRequest pullRequest) {
+
+        final ListenableFuture<ResponseEntity<JsonDocument>> response = getPullRequestInfo(headers, pullRequest, baseUrl);
+        return Async.toSingle(response)
+                .map(entity -> entity.getBody())
+                .map(jsonDocument -> Integer.toString(jsonDocument.read("$.version")));
+
+    }
+
+    private Observable<Card> getCardsForBitbucketServerPR(final HttpHeaders headers,
```

Comment:
I have some ideas for improving the RxJava stuff here - please watch out for a PR requesting to merge into your branch. Basically I'm not a huge fan of returning an `Observable` when we know it only ever emits one item. On a related note, the method name refers to plural cards whereas there's only ever one. (I'll take care of that in my PR as well.)

Reply:
We should create a followup ticket to fix the other connectors after the SingleUtil gets in.  I remember having to switch from single to observable and back a few times for some of this stuff.

Reply:
Please go ahead.

Reply:
I created APF-782 for this.

Reply:
I have taken merged Rob's PR into my branch..